cmd/clusterctl/pkg/client/repository/repository_github.go

```diff
@@ -0,0 +1,287 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package repository
+
+import (
+	"context"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/go-github/github"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/klog"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/config"
+)
+
+const (
+	httpsScheme              = "https"
+	githubDomain             = "github.com"
+	gitHubTokeVariable       = "github-token"
+	gitHubReleaseRepository  = "releases"
+	gitHubLatestReleaseLabel = "latest"
+)
+
+// gitHubRepository provide support for providers hosted on GitHub.
+//
+// We support for GitHub repository that reading from GitHub releases.
+// Repository has full support for version, including also "latest" meta version.
+type gitHubRepository struct {
+	providerConfig           config.Provider
+	configVariablesClient    config.VariablesClient
+	authenticatingHTTPClient *http.Client
+	owner                    string
+	repository               string //this is the GitHub repository
+	defaultVersion           string
+	rootPath                 string
+	componentsPath           string
+	injectClient             *github.Client
+}
+
+var _ Repository = &gitHubRepository{}
+
+func (g *gitHubRepository) DefaultVersion() string {
+	return g.defaultVersion
+}
+
+func (g *gitHubRepository) RootPath() string {
+	return g.rootPath
+}
+
+func (g *gitHubRepository) ComponentsPath() string {
+	return g.componentsPath
+}
+
+func (g *gitHubRepository) GetFile(version, path string) ([]byte, error) {
+	release, err := g.getReleaseByTag(version)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get GitHub release %s", version)
+	}
+
+	// download files from the release
+	files, err := g.downloadFilesFromRelease(release, path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to download files from GitHub release %s", version)
+	}
+
+	return files, nil
+
+}
+
+func newGitHubRepository(providerConfig config.Provider, configVariablesClient config.VariablesClient) (*gitHubRepository, error) {
+	if configVariablesClient == nil {
+		return nil, errors.New("invalid arguments: configVariablesClient can't be nil")
+	}
+
+	rURL, err := url.Parse(providerConfig.URL())
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid url")
+	}
+
+	// if the url is a github repository
```

Comment:
Nit: Use complete sentences for inline comments https://golang.org/doc/effective_go.html#commentary

Include also the reasoning of a code line